libgadget/gravshort-tree.c

```diff
@@ -95,24 +96,32 @@ force_treeev_shortrange(TreeWalkQueryGravShort * input,
 void
 grav_short_tree(const ActiveParticles * act, PetaPM * pm, ForceTree * tree, MyFloat (* AccelStore)[3], double rho0, inttime_t Ti_Current)
 {
-    TreeWalk tw[1] = {{0}};
-    struct GravShortPriv priv;
-    priv.cellsize = tree->BoxSize / pm->Nmesh;
-    priv.Rcut = TreeParams.Rcut * pm->Asmth * priv.cellsize;;
-    priv.G = pm->G;
-    priv.cbrtrho0 = pow(rho0, 1.0 / 3);
-    priv.Ti_Current = Ti_Current;
-    priv.Accel = AccelStore;
+    TreeWalk *tw;
+    cudaMallocManaged(&tw, sizeof(TreeWalk));  // Allocate TreeWalk structure with Unified Memory
+    memset(tw, 0, sizeof(TreeWalk));           // Zero-initialize the structure
+    struct GravShortPriv *priv_ptr;
+    cudaMallocManaged(&priv_ptr, sizeof(struct GravShortPriv));
+
+    // Initialize priv_ptr as usual
+    priv_ptr->cellsize = tree->BoxSize / pm->Nmesh;
+    priv_ptr->Rcut = TreeParams.Rcut * pm->Asmth * priv_ptr->cellsize;
+    priv_ptr->G = pm->G;
+    priv_ptr->cbrtrho0 = pow(rho0, 1.0 / 3);
+    priv_ptr->Ti_Current = Ti_Current;
+    priv_ptr->Accel = AccelStore;
+
     int accelstorealloc = 0;
     if(!AccelStore) {
-        priv.Accel = (MyFloat (*) [3]) mymalloc2("GravAccel", PartManager->NumPart * sizeof(priv.Accel[0]));
+        priv_ptr->Accel = (MyFloat (*) [3]) mymalloc2("GravAccel", PartManager->NumPart * sizeof(priv_ptr->Accel[0]));
         accelstorealloc = 1;
     }
 
     if(!tree->moments_computed_flag)
         endrun(2, "Gravtree called before tree moments computed!\n");
 
-    tw->ev_label = "GRAVTREE";
+    cudaMallocManaged(&tw->ev_label, sizeof(char) * strlen("GRAVTREE") + 1);  // Allocate memory for ev_label
+    strcpy(tw->ev_label, "GRAVTREE");
```

Comment:
Oh no! We can't have static strings on CUDA memory. Can we make it a std::string instead?